File: LogicOverSyntaxCodes/Searching/JavaCodes/Leetcode1283.java
class Solution {
   public static boolean canIPickThisNumber(int[] nums,int x,int threshold){
        int sum=0;
        for (int i = 0; i < nums.length; i++) {
            sum+=(nums[i]+x-1)/x;
            if(sum>threshold){
                return false;
            }
        }
        return true;
    }
    public static int smallestDivisor(int[] nums, int threshold) {
        int max=0;
        for (int i = 0; i < nums.length; i++) {
            max=Math.max(max,nums[i]);
        }
        int start=1,end=max;

        while(start<=end){
            int x=start+(end-start)/2;
            if(canIPickThisNumber(nums,x,threshold)){
                end=x-1;
            }
            else{
                start=x+1;
            }
        }
        return start;
    }
}

File: LogicOverSyntaxCodes/Searching/PythonCodes/Leetcode1283.py
class Solution(object):
    def canIPickThisNumber(self, nums, x, threshold):
        total = 0
        for num in nums:
            total += (num + x - 1) // x
            if total > threshold:
                return False
        return True

    def smallestDivisor(self, nums, threshold):
        max_value = max(nums)
        start, end = 1, max_value

        while start <= end:
            x = start + (end - start) // 2
            if self.canIPickThisNumber(nums, x, threshold):
                end = x - 1
            else:
                start = x + 1

        return start
